lib/usd/translators/shading/usdUVTextureReader.cpp

```diff
@@ -215,42 +220,197 @@ bool PxrMayaUsdUVTexture_Reader::Read(UsdMayaPrimReaderContext* context)
+                depFn.name().asChar());
+            return false;
+        }
+        // Handle UDIM texture files:
+        std::string::size_type udimPos = unresolvedFilePath.rfind(_tokens->UDIMTag.GetString());
+        if (udimPos != std::string::npos) {
+            MPlug tilingAttr = depFn.findPlug(_tokens->uvTilingMode.GetText(), true, &status);
+            if (status == MS::kSuccess) {
+                tilingAttr.setInt(3);
+
+                // USD did not resolve the path to absolute because the file name was not an
+                // actual file on disk. We need to find the first tile to help Maya find the
+                // other ones.
+                std::string udimPath(unresolvedFilePath.substr(0, udimPos));
+                udimPath += "1001";
+                udimPath
+                    += unresolvedFilePath.substr(udimPos + _tokens->UDIMTag.GetString().size());
+
+                Usd_Resolver res(&prim.GetPrimIndex());
+                for (; res.IsValid(); res.NextLayer()) {
+                    std::string resolvedName
+                        = SdfComputeAssetPathRelativeToLayer(res.GetLayer(), udimPath);
+
+                    if (!resolvedName.empty() && !ArIsPackageRelativePath(resolvedName)
+                        && resolvedName != udimPath) {
+                        udimPath = resolvedName;
+                        break;
                     }
-                    val = SdfAssetPath(udimPath);
+                }
+                val = SdfAssetPath(udimPath);
+            }
+        }
+
+        const UsdMayaJobImportArgs& jobArgs = this->_GetArgs().GetJobArguments();
+        if (jobArgs.importUSDZTextures && !jobArgs.importUSDZTexturesFilePath.empty()
+            && !filePath.empty() && ArIsPackageRelativePath(filePath)) {
+            // NOTE: (yliangsiew) Package-relatve path means that we are inside of a USDZ file for
+            // sure...right?
+            // NOTE: (yliangsiew) We use the unresolved file path here that is just
+            // "0/clouds_128_128.png" since that is what `Find()` expects.
+            ArResolver& arResolver = ArGetResolver(); // NOTE: (yliangsiew) This is cached.
+            std::shared_ptr<ArAsset> assetPtr = arResolver.OpenAsset(filePath);
+            if (assetPtr == nullptr) {
+                TF_WARN(
+                    "The file: %s could not be found within the USDZ archive for extraction.",
+                    filePath.c_str());
+                return false;
+            }
+
+            ArAsset*                    asset = assetPtr.get();
+            std::shared_ptr<const char> fileData = asset->GetBuffer();
+            const size_t                fileSize = asset->GetSize();
+
+            bool  needsUniqueFilename = false;
+            char* fileDataBuf = (char*)malloc(sizeof(char) * fileSize);
+            memcpy(fileDataBuf, fileData.get(), fileSize);
+            uint64_t spookyHash = ArchHash64(fileDataBuf, fileSize);
+            free(fileDataBuf);
```

Comment:
Please hash `asset->GetBuffer()` directly without copying it to a temporary buffer.